### DISCUSSION TITLE
Using a DAC that requires a MCLK as an additional input

### DISCUSSION THREAD

**Discussion by mrhempman69** [pschatzmann/ESP32-A2DP#482] [Q&A] [answered]
Hi,

I am trying to use this library with NAU8402. On page 12 of the linked datasheet, it describes that the un-mute sequence for the chip requires "Valid MCLK/FS ratio of 256±3 or 512±3 MCLK cycles per FS has been detected", where "FS" is equivalent to "WS" and is also equivalent to the sample rate.
( https://www.nuvoton.com/export/resource-files/DS_NAU8402_DataSheet_EN_Rev1.7.pdf ). Sure enough, my chip is not outputting any audio, because I don't know how to properly output this kind of information to the MCLK pin. 

In another discussion the maintainer recommended setting "fixed_mclk" and "mclk_io_num" inside i2s_config_t to enable an mclk output. While I can find "fixed_mclk" in the type interface, "mclk_io_num" is not available.

Do I need to use the "audio-tools" library for this? Or is it integrated into this library?

**Comment by mrhempman69** [comment]
I cannot read. It is i2s_**pin**_config_t that has "mck_io_num" (it is not "mclk_io_num" btw, which is as described in the other comment, i thought it was a typo)

**Comment by mrhempman69** [comment]
I did flash the firmware with  mck_io_num set to 23 and fixed_mclk set to true, but there is still no audio output from the DAC. How can I ensure that I have set the MCLK/FS ratio to 256 or 512?

**Comment by pschatzmann** [comment, marked as answer]
But this has nothing to do with my library!

Wouldn't you read the ESP32 documentation first? https://docs.espressif.com/projects/esp-idf/en/v4.2.2/esp32/api-reference/peripherals/i2s.html

That's complete nonsense: you can't use pin 23 for this and setting fixed_mclk to true does not make any sense either!

You need to provide the frequency to fixed_mclk and if I remember right only GPIO00, GPIO01 and GPIO02 can be used as master clock pins. But you better double check with the documentation.

ps. If you activated the logging you should have seen the error message that 23 is not supported!

**Comment by mrhempman69** [comment]
I have found out that I need to use ESP-IDF 5.1.1, because it contains "I2S_MULTIPLE_MCLK_256" for exactly this purpose. However, I do not think ESP-IDF 5.1.1 is used by arduino or platformio, so I'm following the instructions here:

https://github.com/pschatzmann/ESP32-A2DP/wiki/Espressif-IDF-as-a-Component
 so that I can use the latest ESP-IDF 5.1.1 release...

For some reason, the "idf.py create_project esp32-adp" command doesn't work, but that's not this library's problem... I got around this by creating a new project using the VSCode ESP-IDF extension, and continued with the rest of your guide.

Everything works up until "idf.py build".... The compiler doesn't like that your library wants to do `printf` with %d for "long unsigned int"....  It is a "-Werror=format=". However, I think even if I were to tell the compiler to ignore this error (not sure how yet), it would still hit these errors:
```

C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:281:28: error: 'PIN_CTRL' was not declared in this scope
  281 |             WRITE_PERI_REG(PIN_CTRL, 0xFF00);

C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:280:29: error: 'PERIPHS_IO_MUX_U0RXD_U' was not declared in this scope
  280 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0RXD_U, FUNC_U0RXD_CLK_OUT2);
      |                             ^~~~~~~~~~~~~~~~~~~~~~
C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:280:53: error: 'FUNC_U0RXD_CLK_OUT2' was not declared in this scope
  280 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0RXD_U, FUNC_U0RXD_CLK_OUT2);

C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:276:29: error: 'PERIPHS_IO_MUX_U0TXD_U' was not declared in this scope
  276 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0TXD_U, FUNC_U0TXD_CLK_OUT3);
      |                             ^~~~~~~~~~~~~~~~~~~~~~
C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:276:53: error: 'FUNC_U0TXD_CLK_OUT3' was not declared in this scope
  276 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0TXD_U, FUNC_U0TXD_CLK_OUT3);

C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:272:29: error: 'PERIPHS_IO_MUX_GPIO0_U' was not declared in this scope
  272 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);
      |                             ^~~~~~~~~~~~~~~~~~~~~~
C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:272:53: error: 'FUNC_GPIO0_CLK_OUT1' was not declared in this scope
  272 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);
      |                                                     ^~~~~~~~~~~~~~~~~~~
C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:272:13: error: 'PIN_FUNC_SELECT' was not declared in this scope
  272 |             PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);

C:/Users/mrhem/OneDrive/Documents/esp32-a2dp-espidf3/template-app/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:47:30: error: 'I2S_MCLK_MULTIPLE_DEFAULT' was not declared in this scope; did you mean 'I2S_MCLK_MULTIPLE_512'?
   47 |             .mclk_multiple = I2S_MCLK_MULTIPLE_DEFAULT,
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~




```
Along with a few other similar errors (PIN_CTRL especially)...

I tried both the most recent release (1.7.4) and the master branch. Both hit these same errors...

**Comment by pschatzmann** [comment]
But your assumption is completely wrong!
In the actual Arduino Releases you can simply use mck_io_num as you have done above and in older releases you could use the  i2s_mclk_pin_select() method provided by the BluetoothA2DPSink.

In any case you are limited to the supported pins mentioned above!

If you want to use 5.1.1 I suggest that you deactivate the I2S output in this project and use the I2S output of the AudioTools which should support both the old and new I2S APIs.

ps. I committed a correction to exclude the i2s_mclk_pin_select() method starting from 5.1.1

**Reply by mrhempman69** [reply]
Ah, I just realized that some of my math for setting the MCLK frequency was wrong when I was trying with the 4.2.2 library that you linked to.  However, I also haven't found anywhere that the documentation lists the few limited GPIOs that you mentioned, so I have been trying with GPIO01 just in case... I will try everything mentioned here tomorrow

**Reply by pschatzmann** [reply]
Have a look at the i2s_mclk_pin_select() method! There you can find the pins as well: 0, 1 and 3
This is documented in the Technical Reference Manual.
